rfcs/0000-trust-store-for-tls-and-ssh/README.md

```diff
@@ -0,0 +1,344 @@
+- Formalize support for configuring trust at controller-level.
+- Add toggle to block object-level trust overrides.
+- Enable users to surface trust information securely.
+
+### Non-Goals
+
+- Maintain backwards compatibility with older versions of Flux.
+
+## Proposal
+
+For configuring system-wide trust, Flux would rely on the well-established OS-level
+trust stores. When dynamically mounting of the trust store is required, it will be
+enabled by using Kubernetes `Secret` and `ConfigMap` mounting. When immutable trust
+store is required, users can build their own version of the controllers, with their
+baked-in settings.
+
+TLS and SSH use different techniques to establish the identity of remote servers,
+each relying on its own trust store.
+
+The sections below will dive into the specifics of each one, highlighting their
+details, changes required and example of the proposed usage.
+
+A new way to configure object-level Trust Store overrides is also being proposed,
+in combination with a controller level toggle to disable it.
+
+### SSH
+
+In SSH, the remote server identity is based on [Trust on first use]. At first
+connection to a new server, the user confirms whether or not to trust that server
+based on the server's Public key fingerprint.
+
+In the context of Flux, which provides no user interaction, if the remote server
+finger print is not configured within the provided set of Known Hosts, the
+connection is aborted.
+
+#### Controller-level Known Hosts
+
+For setting controller level Known Hosts, we propose the use of the existing
+Linux file in disk: [/etc/ssh/ssh_known_hosts].
+
+Users would be able to configure the OS level trust store by mounting either
+a `ConfigMap` or `Secret` directly into the Flux Controllers.
+
+`ConfigMap` example:
+```yaml
+apiVersion: v1
+kind: ConfigMap
+metadata:
+  name: flux-trust-store
+  namespace: flux-system
+data:
+  known_hosts: |
+    github.com ecdsa-sha2-nistp256 AAAAE2VjZHNhLXNoYTItbmlzdHAyNTYAAAAIbmlzdHAyNTYAAABBBEmKSENjQEezOmxkZMy7opKgwFB9nkt5YRrYMjNuG5N87uRgg6CLrbo5wAdT/y6v0mKV0U2w0WZ2YB/++Tpockg=
+```
+
+Patch required on the main `kustomization.yaml`:
+```yaml
+  - patch: |
+      - op: add
+        path: /spec/template/spec/containers/0/volumeMounts
+        value:
+          - name: ssh-trust-store
+            mountPath: /etc/ssh/ssh_known_hosts
+            subPath: known_hosts
+            readOnly: true
+      - op: add
+        path: /spec/template/spec/volumes
+        value:
+          - name: ssh-trust-store
+            configMap:
+              name: flux-trust-store
+    target:
+      kind: Deployment
+      name: "(kustomize|image-automation|source|image-reflector|helm|notification)-controller"
+```
+
+#### Object-level Known Hosts Expansion
+
+A new field is to be introduced into the existing kinds `ImageUpdateAutomation` and
+`GitRepository`, to allow users to expand on the controller-level known hosts for
+SSH operations:
```

Comment:
I find this very problematic, why would we allow for a namespaced object own by a tenant to alter the trust store for all tenants? This also makes Flux bootstrap break, since you can't control which CR is reconciled first, if people add a trust store in one CR we can't reconcile that first to make it available to others.

Reply:
It does, it expands the trust store for that object only.

Reply:
How does this cope with the know_hosts from the `secretRef`? Can you explain the differences in the doc please?

Reply:
I will rephrase to make that clearer. But basically that's what we already do at present based on the CA bundles provided by secrets, however this would be based on a field instead.